crates/uniswapx-rs/src/order.rs

```diff
@@ -285,6 +400,131 @@ fn resolve_decay(
     }
 }
 
+impl NonlinearDutchDecay {
+
+    /// Calculates the decayed amount based on the current block and the defined curve
+    pub fn decay(
+        &self,
+        start_amount: Uint<256, 4>,
+        decay_start_block: Uint<256, 4>,
+        current_block: Uint<256, 4>,
+        min_amount: Uint<256, 4>,
+        max_amount: Uint<256, 4>,
+    ) -> Result<Uint<256, 4>, Box<dyn Error>> {
+        // Check for invalid decay curve
+        if self.relativeAmounts.len() > 16 {
+            return Err("Invalid decay curve".into());
+        }
+
+        // Handle current block before decay or no decay
+        if decay_start_block >= current_block || self.relativeAmounts.is_empty() {
+            return Ok(start_amount.clamp(min_amount, max_amount));
+        }
+
+        // Cap block_delta to u16::MAX to prevent overflow
+        let block_delta: u16 = u16::try_from(
+            (current_block - decay_start_block).min(Uint::<256, 4>::from(u16::MAX))
+        )?;
+        println!("block_delta: {}", block_delta);
+
+        let (start_point, end_point, rel_start_amount, rel_end_amount) = 
+            self.locate_curve_position(block_delta)?;
+        println!("start_point: {}", start_point);
+        println!("end_point: {}", end_point);
+        println!("rel_start_amount: {}", rel_start_amount);
+        println!("rel_end_amount: {}", rel_end_amount);
+
+        // Calculate decay of only the relative amounts
+        let curve_delta = resolve_decay(
+            Uint::<256, 4>::from(block_delta),
+            Uint::<256, 4>::from(start_point),
+            Uint::<256, 4>::from(end_point),
+            Uint::<256, 4>::try_from(rel_start_amount).unwrap(),
+            Uint::<256, 4>::try_from(rel_end_amount).unwrap(),
```

Comment:
I know we used `unwrap()` in in the existing code 😅 but generally should be avoided, e.g. can do `.expect("<what went wrong")` instead to make debugging easier